src/boosting/gbdt_prediction.cpp

```diff
@@ -17,44 +18,33 @@
 
 namespace LightGBM {
 
-void GBDT::PredictRaw(const double* features, double* output) const {
-  if (num_threads_ <= num_tree_per_iteration_) {
-    #pragma omp parallel for schedule(static)
-    for (int k = 0; k < num_tree_per_iteration_; ++k) {
-      for (int i = 0; i < num_iteration_for_pred_; ++i) {
-        output[k] += models_[i * num_tree_per_iteration_ + k]->Predict(features);
-      }
-    }
-  } else {
-    for (int k = 0; k < num_tree_per_iteration_; ++k) {
-      double t = 0.0f;
-      #pragma omp parallel for schedule(static) reduction(+:t)
-      for (int i = 0; i < num_iteration_for_pred_; ++i) {
-        t += models_[i * num_tree_per_iteration_ + k]->Predict(features);
-      }
-      output[k] = t;
-    }
+void GBDT::PredictRaw(const double* features, double* output, const PredictionEarlyStopInstance* earlyStop) const {
+  const auto noEarlyStop = createPredictionEarlyStopInstance("none", PredictionEarlyStopConfig());
+  if (earlyStop == nullptr)
+  {
+    earlyStop = &noEarlyStop;
   }
-}
 
-void GBDT::Predict(const double* features, double* output) const {
-  if (num_threads_ <= num_tree_per_iteration_) {
-    #pragma omp parallel for schedule(static)
+  int earlyStopRoundCounter = 0;
+  for (int i = 0; i < num_iteration_for_pred_; ++i) {
+    // predict all the trees for one iteration
     for (int k = 0; k < num_tree_per_iteration_; ++k) {
-      for (int i = 0; i < num_iteration_for_pred_; ++i) {
-        output[k] += models_[i * num_tree_per_iteration_ + k]->Predict(features);
-      }
+      output[k] += models_[i * num_tree_per_iteration_ + k]->Predict(features);
     }
-  } else {
-    for (int k = 0; k < num_tree_per_iteration_; ++k) {
-      double t = 0.0f;
-      #pragma omp parallel for schedule(static) reduction(+:t)
-      for (int i = 0; i < num_iteration_for_pred_; ++i) {
-        t += models_[i * num_tree_per_iteration_ + k]->Predict(features);
-      }
-      output[k] = t;
+
+    // check early stopping
+    ++earlyStopRoundCounter;
+    if (earlyStop->roundPeriod == earlyStopRoundCounter) {
+      if (earlyStop->callbackFunction(output, num_tree_per_iteration_))
+        return;
+      earlyStopRoundCounter = 0;
     }
   }
+}
+
+void GBDT::Predict(const double* features, double* output, const PredictionEarlyStopInstance* earlyStop) const {
+  PredictRaw(features, output, earlyStop);
```

Comment:
I will try. Anyway this call is probably not much compared to predicting each individual tree.